libyogrt ^mpich %oneapi fails build: configure: error: slurm is not in specified location!

### Steps to reproduce the issue

`libyogrt ^mpich %oneapi` fails to build using:
* spack@develop ( from ``)
* Ubuntu 20.04, X86_64
* GCC 11.4.0
* Docker container image `ecpe4s/ubuntu20.04-runner-amd64-oneapi-2023.2.1:2023.08.01`

<details><summary>Spack environment</summary>

```
spack:
  view: false

  concretizer:
    reuse: false
    unify: false

  compilers:
  - compiler:
      spec: oneapi@2023.2.1
      paths:
        cc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icx
        cxx: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icpx
        f77: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
        fc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
      flags: {}
      operating_system: ubuntu20.04
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []
  - compiler:
      spec: gcc@=11.4.0
      paths:
        cc: /usr/bin/gcc
        cxx: /usr/bin/g++
        f77: /usr/bin/gfortran
        fc: /usr/bin/gfortran
      flags: {}
      operating_system: ubuntu20.04
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []

  packages:
    all:
      require: '%oneapi'
      providers:
        blas: [openblas]
        mpi: [mpich]
        tbb: [intel-tbb]
      target: [x86_64]
      variants: +mpi
    elfutils:
      variants: +bzip2 ~nls +xz
    hdf5:
      require: "%gcc"
      variants: +fortran +hl +shared
    libfabric:
      variants: fabrics=sockets,tcp,udp,rxm
    libunwind:
      variants: +pic +xz
    ncurses:
      variants: +termlib
    openblas:
      variants: threads=openmp
    trilinos:
      variants: +amesos +amesos2 +anasazi +aztec +belos +boost +epetra +epetraext
        +ifpack +ifpack2 +intrepid +intrepid2 +isorropia +kokkos +ml +minitensor +muelu
        +nox +piro +phalanx +rol +rythmos +sacado +stk +shards +shylu +stokhos +stratimikos
        +teko +tempus +tpetra +trilinoscouplings +zoltan +zoltan2 +superlu-dist gotype=long_long
    xz:
      variants: +pic
    mesa:
      version: [21.3.8]
    mpi:
      require: mpich
    mpich:
      require: '@4.1.2 ~wrapperrpath ~hwloc'
    py-cryptography:
      require: '@38.0.1'
    unzip:
      require: '%gcc'
    binutils:
      require: '%gcc'
      variants: +ld +gold +headers +libiberty ~nls
    llvm:
      require: '%gcc'
    ruby:
      require: '%gcc'
    rust:
      require: '%gcc'
    krb5:
      require: '%gcc'
    papi:
      require: '%gcc'
    openssh:
      require: '%gcc'
    libffi:
      require: "@3.4.4"
    dyninst:
      require: "%gcc"
    bison:
      require: '%gcc'

  specs:
  - scr               # libyogrt: configure: error: slurm is not in specified location!
```

</pre></details>


<details><summary>Concretization</summary>

<pre>
==> Environment concretized in 17.99 seconds.
==> Concretized scr
 -   4wjrntu  scr@3.0.1%oneapi@2023.2.1+bbapi~bbapi_fallback~dw+examples+fortran~ipo+libyogrt+pdsh+shared+tests build_system=cmake build_type=Release cache_base=/dev/shm cntl_base=/dev/shm copy_config=none file_lock=FLOCK generator=make resource_manager=SLURM scr_config=scr.conf arch=linux-ubuntu20.04-x86_64
[+]  jy456tl      ^axl@0.7.1%oneapi@2023.2.1+bbapi~bbapi_fallback~dw~ipo+pthreads+shared async_api=daemon build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  u5lxz2v      ^cmake@3.26.3%oneapi@2023.2.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
[+]  x75zxog          ^curl@8.1.2%oneapi@2023.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-x86_64
[+]  cqdtgb2              ^nghttp2@1.52.0%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  fbsztq2              ^openssl@3.1.2%oneapi@2023.2.1~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
[+]  kylhswq                  ^ca-certificates-mozilla@2023-05-30%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  jux6zrw                  ^perl@5.38.0%oneapi@2023.2.1+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  ezmplcp                      ^berkeley-db@18.1.40%oneapi@2023.2.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
[+]  pvdjart          ^ncurses@6.4%oneapi@2023.2.1~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  tns2r7q      ^dtcmp@1.1.4%oneapi@2023.2.1+shared build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  6lyanav          ^lwgrp@1.0.5%oneapi@2023.2.1+shared build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  zniny5g      ^er@0.2.0%oneapi@2023.2.1~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  wnkqq3t      ^gmake@4.4.1%oneapi@2023.2.1~guile build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  57cuyfr      ^kvtree@1.3.0%oneapi@2023.2.1~ipo+mpi+shared build_system=cmake build_type=Release file_lock=FLOCK generator=make arch=linux-ubuntu20.04-x86_64
 -   wwouk7q      ^libyogrt@1.33%oneapi@2023.2.1~static build_system=autotools scheduler=slurm arch=linux-ubuntu20.04-x86_64
[+]  v6b7rml          ^slurm@23-02-2-1%oneapi@2023.2.1~gtk~hdf5~hwloc~mariadb~pmix+readline~restd build_system=autotools sysconfdir=PREFIX/etc arch=linux-ubuntu20.04-x86_64
[+]  kzn5ow6              ^glib@2.76.4%oneapi@2023.2.1~libmount~strip build_system=meson buildtype=release default_library=shared tracing=none arch=linux-ubuntu20.04-x86_64
[+]  f3vlv6j                  ^elfutils@0.189%oneapi@2023.2.1~debuginfod~nls build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  prmjhbw                      ^zstd@1.5.5%oneapi@2023.2.1+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  jkwp5fn                  ^gettext@0.21.1%oneapi@2023.2.1+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  asawgkg                      ^tar@1.34%oneapi@2023.2.1 build_system=autotools zip=pigz arch=linux-ubuntu20.04-x86_64
[+]  7kmvh6m                          ^pigz@2.7%oneapi@2023.2.1 build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]  njghyz5                  ^libffi@3.4.4%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  vcxonba                  ^meson@1.2.0%oneapi@2023.2.1 build_system=python_pip patches=0f0b1bd arch=linux-ubuntu20.04-x86_64
[+]  rhtdjg7                      ^py-pip@23.1.2%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  r27dsjk                      ^py-setuptools@68.0.0%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  3e76ad3                      ^py-wheel@0.37.1%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  dmuvqoy                  ^ninja@1.11.1%oneapi@2023.2.1+re2c build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  u7padth                      ^re2c@2.2%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  uilkym5                  ^pcre2@10.42%oneapi@2023.2.1~jit+multibyte build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  6u3stx4              ^json-c@0.16%oneapi@2023.2.1~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  6rkfqlz              ^lz4@1.9.4%oneapi@2023.2.1 build_system=makefile libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  dulak4n              ^munge@0.5.15%oneapi@2023.2.1 build_system=autotools localstatedir=PREFIX/var arch=linux-ubuntu20.04-x86_64
[+]  vohypiq                  ^bzip2@1.0.8%oneapi@2023.2.1~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  toxim4n                  ^libgcrypt@1.10.2%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  3y4rvwc                      ^libgpg-error@1.47%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  lk7ij2z                          ^gawk@5.2.1%oneapi@2023.2.1~nls build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  eal5xjd                              ^gmp@6.2.1%oneapi@2023.2.1+cxx build_system=autotools libs=shared,static patches=69ad2e2 arch=linux-ubuntu20.04-x86_64
[+]  azdxia3                              ^mpfr@4.2.0%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  tikhr5m                                  ^autoconf-archive@2023.02.20%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  kv6w3hr                                  ^texinfo@7.0.3%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  d5ulnqr              ^readline@8.2%oneapi@2023.2.1 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
[+]  ypccx6c      ^mpich@4.1.2%oneapi@2023.2.1~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]  exj2sv2          ^findutils@4.9.0%oneapi@2023.2.1 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-x86_64
[+]  auek3j2          ^libfabric@1.18.1%oneapi@2023.2.1~debug~kdreg build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
[+]  wrjo7ui          ^libpciaccess@0.17%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  znew2ad              ^libtool@2.4.7%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  kntnrq7              ^util-macros@1.19.3%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  ycyu5fa          ^libxml2@2.10.3%oneapi@2023.2.1~python build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  d4jkpaa              ^libiconv@1.17%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  ypt5s7x              ^xz@5.4.1%oneapi@2023.2.1+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  tjpjhh6          ^pkgconf@1.9.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  vczl2ng          ^yaksa@0.2%oneapi@2023.2.1~cuda~rocm build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  jush4ya              ^autoconf@2.69%oneapi@2023.2.1 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
[+]  zo6xfq7              ^automake@1.16.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  nucb3bv              ^m4@1.4.19%oneapi@2023.2.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]  mkd2zhf                  ^diffutils@3.9%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  k73plmc                  ^libsigsegv@2.14%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  glzlgbk              ^python@3.8.17%oneapi@2023.2.1+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]  ozz5pat                  ^expat@2.5.0%oneapi@2023.2.1+libbsd build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  43lzrfm                      ^libbsd@0.11.7%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  a7vy5xz                          ^libmd@1.0.4%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  isd3hgh                  ^gdbm@1.23%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  ly2g5ah                  ^libxcrypt@4.4.35%oneapi@2023.2.1~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu20.04-x86_64
[+]  cvf6jry                  ^sqlite@3.42.0%oneapi@2023.2.1+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  fbgrl26                  ^util-linux-uuid@2.38.1%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  depzymz      ^pdsh@2.31%oneapi@2023.2.1+ssh+static_modules build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  dyi3u6t      ^rankstr@0.1.0%oneapi@2023.2.1~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  qcg3jid      ^redset@0.2.0%oneapi@2023.2.1~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  kr6lqlu      ^shuffile@0.2.0%oneapi@2023.2.1~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  mn665p5      ^spath@0.2.0%oneapi@2023.2.1~ipo+mpi+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  cvtzwql      ^zlib@1.2.13%oneapi@2023.2.1+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64
</pre></details>

Error:
```
$> spack -e . install
...
==> Installing libyogrt-1.33-wwouk7qlys2km355l4zjwrkqknc75snj [52/53]
==> No binary for libyogrt-1.33-wwouk7qlys2km355l4zjwrkqknc75snj found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/79/797d20c49cdc4f6beae8660b4f41ba7ac13f7e93a0344b47f0bdc64f780d1398.tar.gz
==> No patches needed for libyogrt
==> libyogrt: Executing phase: 'autoreconf'
==> libyogrt: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/root/spack-stage/spack-stage-libyogrt-1.33-wwouk7qlys2km355l4zjwrkqknc75snj/spack-src/configure' '--prefix=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2023.2.1/libyogrt-1.33-wwouk7qlys2km355l4zjwrkqknc75snj' '--with-slurm=/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2023.2.1/slurm-23-02-2-1-v6b7rmlxse5pjkdwjnun6irpuchaapbq'

1 error found in build log:
     83    checking whether to build shared libraries... yes
     84    checking whether to build static libraries... no
     85    checking for pkg-config... /usr/bin/pkg-config
     86    checking pkg-config is at least version 0.9.0... yes
     87    checking for inline... inline
     88    checking for slurm installation...
  >> 89    configure: error: slurm is not in specified location!

See build log for details:
  /tmp/root/spack-stage/spack-stage-libyogrt-1.33-wwouk7qlys2km355l4zjwrkqknc75snj/spack-build-out.txt
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (00f8f5898a8f81605d733bcc9296628a10d3aa31)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake_avx512
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/12360666/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/12360667/spack-build-out.txt)
[config.log](https://github.com/spack/spack/files/12360668/config.log)

@CamStan @gonsie @robertkb @adammoody 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

From the ``configure.out`` file, libyogrt's configure is failing when trying to compile a file with ``icx`` while checking for ``slurm_get_rem_time()``: 

```
configure:8998: checking for slurm installation
configure:9025: /spack/lib/spack/env/oneapi/icx -o conftest -O2   conftest.c -L/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2023.2.1/slurm-23-02-2-1-v6b7rmlxse5pjkdwjnun6irpuchaapbq/lib -lslurm -lpthread -lcrypto  >&5
conftest.c:30:1: error: call to undeclared function 'slurm_get_rem_time'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
slurm_get_rem_time(0);
^
1 error generated.
configure:9025: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "Your One Get Remaining Time library"
| #define PACKAGE_TARNAME "libyogrt"
| #define PACKAGE_VERSION "1.33"
| #define PACKAGE_STRING "Your One Get Remaining Time library 1.33"
| #define PACKAGE_BUGREPORT "https://github.com/LLNL/libyogrt/issues"
| #define PACKAGE_URL "https://github.com/LLNL/libyogrt"
| #define PACKAGE "libyogrt"
| #define VERSION "1.33"
| #define STDC_HEADERS 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_SYS_STAT_H 1
| #define HAVE_STDLIB_H 1
| #define HAVE_STRING_H 1
| #define HAVE_MEMORY_H 1
| #define HAVE_STRINGS_H 1
| #define HAVE_INTTYPES_H 1
| #define HAVE_STDINT_H 1
| #define HAVE_UNISTD_H 1
| #define HAVE_DLFCN_H 1
| #define LT_OBJDIR ".libs/"
| #define LIBYOGRT_LT_CURRENT /**/
| #define LIBYOGRT_LT_REVISION /**/
| #define LIBYOGRT_LT_AGE /**/
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| slurm_get_rem_time(0);
|   ;
|   return 0;
| }
configure:9039: result: 
configure:9054: error: slurm is not in specified location!
```

That comes from this test:

https://github.com/LLNL/libyogrt/blob/4346565dd36e1b847206dc35de1e65bd787a3473/config/x_ac_slurm.m4#L52-L56

Perhaps the libyogrt configure test can be updated to include the proper SLURM header, as done with FLUX?

https://github.com/LLNL/libyogrt/blob/4346565dd36e1b847206dc35de1e65bd787a3473/config/x_ac_flux.m4#L51

In the short term, perhaps ``-Wimplicit-function-declaration`` could be passed to the icx compiler to disable error.

Does it help if you add this to the ``flags:`` entry of the oneapi section in the spack environment?

https://spack-tutorial.readthedocs.io/en/latest/tutorial_configuration.html#compiler-flags

```
  compilers:
  - compiler:
      spec: oneapi@2023.2.1
      paths:
        cc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icx
        cxx: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icpx
        f77: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
        fc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
      flags:
        cppflags: -Wimplicit-function-declaration
        cflags: -Wimplicit-function-declaration
```

--- a/var/spack/repos/builtin/packages/libyogrt/package.py
+++ b/var/spack/repos/builtin/packages/libyogrt/package.py
@@ -61,6 +61,12 @@ class Libyogrt(AutotoolsPackage):
                 version
             )
 
+    def flag_handler(self, name, flags):
+        if name == "cflags":
+            if self.spec.satisfies("%oneapi"):
+                flags.append("-Wno-error=implicit-function-declaration")
+        return (flags, None, None)
+
     def configure_args(self):
         args = []
 

